blog: new post on locators (#417)
debs-obrien/debbie.codes: content/blog/how-to-locate-elements-in-playwright.md

date: 2022-11-30
description: How can we locate a link click on it in our e2e tests. In this article we will learn how to locate elements in Playwright using the test generator to generate your tets based on user actions or pick a locator by clicking on an element
image: v1669817488/debbie.codes/blog/2022/vs-code-codegen_gxf58x.png
alt: browser window showing highlighted element and inspector window with explore showing the locator highlighted
ogImage: https://res.cloudinary.com/debsobrien/image/upload/v1669817488/debbie.codes/blog/2022/vs-code-codegen_gxf58x.png
provider: cloudinary
tags: [testing, playwright]
published: true
---

In order to write end to end tests we need to first find elements on the webpage and then perform user actions on them. For example, find a link and click on it. 

>The most convenient way to find an element is to use [Playwright's test generator](https://playwright.dev/docs/codegen-intro), which allows you to not only generate tests but also generate a locator for a specific element.

But before we dive into how to use the test generator, first let's understand what a link element is and what exactly locators are.

## What is a link?

> The HTML [`<a>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), also known as the anchor element, is an interactive element that creates a hyperlink to another page either within your site or externally to another site. It can also be use to link to specific areas on the same page as well as for emails, file downloads or anything else a URL can address. The `<a>` element uses the `href` attribute to pass in the URL that the hyperlink points to.

Reviewer: ```suggestion
The HTML [`<a>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), also known as the anchor element, is an interactive element that creates a hyperlink to another page either within your site or externally to another site. It can also be used to link to specific areas on the same page as well as for emails, file downloads or anything else a URL can address. The `<a>` element uses the `href` attribute to pass in the URL that the hyperlink points to.
```

```js
await page.getByRole('link', { name: /read more about locators/i }).click();
```

When you are writing your tests it can be quite difficult to know which accessible role to use if using the `getByRole()` locator. For a link it is pretty straight forward but there is a vast list of roles that can be used. If your element doesn't have an accessible locator then choosing the best locator to use is not easy especially if you are not familiar with the DOM structure of the site you are testing. So how can you choose the best locator? 

>Don't spend time trying to figure out which is the best locator to use. Instead use the test generator to generate your locator.

---

## Playwright's test generator

You can generate code right from VS Code when using the [VS Code extension](https://playwright.dev/docs/getting-started-vscode#installation) by clicking the 'record new' link from the testing sidebar. 

![test generator in vs code editor](https://res.cloudinary.com/debsobrien/image/upload/v1669817488/debbie.codes/blog/2022/vs-code-codegen_gxf58x.png)

> As you click on an element on your website, the generator will look at your DOM and figure out the best locator for you, prioritizing role based locators over other locators such as text or test id locators. If the generator comes across two locators with the same name it will either chain locators or use the [`filter()`](https://playwright.dev/docs/locators#filtering-locators) method to ensure the locator is resilient so you don't have to worry about failing tests due to locators.

Reviewer: ```suggestion
As you click on an element on your website, the generator will look at your page and figure out the best locator for you, prioritizing role, text and test id locators. If the generator finds multiple elements matching the locator, it will improve the locator to make it resilient and uniquely identify the target element, so you don't have to worry about failing tests due to locators. For example, it could chain locators to narrow down the scope, or use the [`filter()`](https://playwright.dev/docs/locators#filtering-locators) method to select between multiple elements.
```

> If you are not using the VS Code extension Playwright comes with a standalone test generator called [Codegen](https://playwright.dev/docs/test generator-intro).

Reviewer: ```suggestion
If you are not using the VS Code extension, Playwright comes with a standalone test generator called [Codegen](https://playwright.dev/docs/test generator-intro).
```

```bash
npx playwright codegen
```

![Codegen](https://res.cloudinary.com/debsobrien/image/upload/v1669647195/debbie.codes/blog/2022/codegen_qykizt.png)

### Pick a locator

You can pick a locator by clicking the [pick locator](https://playwright.dev/docs/next/getting-started-vscode#picking-a-locator) button from the testing sidebar when using the VS Code Extension. 

![pick locator from vs code highlighting the locator in browser window](https://res.cloudinary.com/debsobrien/image/upload/v1669817100/debbie.codes/blog/2022/pick-locator-vscode_xrcdkc.png)

When you hover over any element in the browser window the locator for that element will be highlighted under your cursor. If you click on the element the locator will appear in the 'pick locator' box. You can then copy it to your clipboard and paste it into your test.

You can also pick a locator when using the standalone test generator by clicking the explore button.

```bash
npx playwright open
```

![test generator generating a getByRole locator for a link](https://res.cloudinary.com/debsobrien/image/upload/v1669795104/debbie.codes/blog/2022/locator-link_hz7icx.png)

### Chaining locators

> If your site contains duplicate links, for example two about links, one in the header and one in the footer, then the test generator will create a unique locator for each one. It does this by looking at the next accessible ancestor to the link and chains the locators.

Reviewer: ```suggestion
If your site contains duplicate links, for example two about links, one in the header and one in the footer, then the test generator will create a unique locator for each one. It does this by looking for an easily-identifiable ancestor of the link and chains the locators.
```

```js
getByRole('navigation').getByRole('link', { name: 'About' })
```

![test generator highlighting the about link in the header](https://res.cloudinary.com/debsobrien/image/upload/v1669795104/debbie.codes/blog/2022/locator-chaining-header_hacas1.png)

> For the about link in the footer the test generator again looks at the next accessible ancestor to the link and chains the two [`getByRole()`](https://playwright.dev/docs/locators#locate-by-role) locators ensuring that there is only one instance for this locator.

Reviewer: ```suggestion
For the about link in the footer the test generator again looks for an ancestor and chains the two [`getByRole()`](https://playwright.dev/docs/locators#locate-by-role) locators ensuring that there is only one element matching this locator.
```

```js
getByRole('contentinfo').getByRole('link', { name: 'About' })
```

![test generator highlighting the footer link](https://res.cloudinary.com/debsobrien/image/upload/v1669795104/debbie.codes/blog/2022/locator-chaining-footer_ogczv7.png)

### Filtering locators

> If the test generator can't give you a unique locator when chaining it will use the [`filter()`](https://playwright.dev/docs/locators#filtering-locators) method to ensure the locator is unique. For example we might have two lists that contain topics for blog posts. The first list is used to filter the blog posts and the second list is used to display the topics for each post. The test generator will use the [`filter()`](https://playwright.dev/docs/locators#filtering-locators) method to ensure the locator is unique.

Reviewer: ```suggestion
If the test generator can't give you a unique locator through chaining it will use the [`filter()`](https://playwright.dev/docs/locators#filtering-locators) method to ensure the locator is unique. For example, we might have two lists that contain topics for blog posts. The first list is used to filter the blog posts and the second list is used to display the topics for each post. The test generator will use the [`filter()`](https://playwright.dev/docs/locators#filtering-locators) method to ensure the locator is unique.
```

```js
getByRole('list')
  .filter({ hasText: 'architecturedev reljamstackjavascriptlifestylementoringmotivationnuxtperformance' })
  .getByRole('link', { name: 'architecture' })
```

![test generator generating a list locator with filter method](https://res.cloudinary.com/debsobrien/image/upload/v1669795104/debbie.codes/blog/2022/locator-filtering_hvlbco.png)

### Improving your locators

> You may notice that filtering by text can sometimes give you a really ugly locator, like in the example above, and may cause you issues later, especially if you add another topic to the filters. If you want to improve this locator you first need to improve the HTML. For example you can add an aria-label to the first `<ul>` element with the value of topics. This not only helps improve the accessability of the page but also allows Playwright to locate by a role selector.

Reviewer: I think we should first explain that you can tweak locator to make it better in this case. For example, replace long `hasText` with shorter one that is resilient. And only as a last resort, you would go and change the page itself.

> You may notice that filtering by text can sometimes give you a really ugly locator, like in the example above, and may cause you issues later, especially if you add another topic to the filters. If you want to improve this locator you first need to improve the HTML. For example you can add an aria-label to the first `<ul>` element with the value of topics. This not only helps improve the accessability of the page but also allows Playwright to locate by a role selector.

Reviewer: ```suggestion
You may notice that filtering by text can sometimes give you a really ugly locator, like in the example above, and may cause you issues later, especially if you add another topic to the filters. If you want to improve this locator you first need to improve the HTML. For example you can add an aria-label to the first `<ul>` element with the value of topics. This not only helps improve the accessibility of the page but also allows Playwright to locate by role.
```
